050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,801 @@
+    ```yaml
+    status:
+      conditions:
+      - type: Ready
+        status: False
+        reason: KafkaError
+        message: Deletion failed: ${Error_Message}
+        lastTransitionTime: 20230301T103000Z
+    ```
+
+The presence of this finalizer will be checked (and added, if missing) on every reconciliation where the `metadata.deletionTimestamp` isn't set. 
+
+---
+
+#### Example 3: `KafkaTopic` deletion propagates to Kafka
+
+> 1. The topic is currently Kube-managed
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 123
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       # ...
+>     status:
+>       observedGeneration: 123 # == generation => it is Kube-managed
+>     ```
+> 
+> 2. The user then deletes the resource (e.g. `kubectl delete kafkatopic foo`).
+>     Because of the presence of a `metadata.finalizer` the resource remains present.
+>     Instead Kube adds the `metadata.deletionTimestamp` field.
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>       deletionTimestamp: 20230301T000000.000
+>     spec:
+>       # ...
+>     status:
+>       observedGeneration: 124 # == generation => it is not syncrhonized
+>     ```
+> 
+> 3. The operator notices the `metadata.deletionTimestamp` field.
+>    1. If there is more than one `KafkaTopic` which is trying to manage this topic then deletion from Kafka is not attempted;
+>       the operator will remove the finalizer and Kube will remove the resource; 
+>       upon timed reconciliation the condition on the other `KafkaTopic` will be removed and reconciliation via that other `KafkaTopic` will proceed as normal.
+>    2. If deletion is disallowed via `delete.topic.enable=false` broker configuration then the operator will remove the finalizer and Kube will remove the resource. 
+>       The topic in Kafka will thus become unmanaged.
+>    3. Otherwise, the operator attempts to deletes the topic from Kafka.
+>        1. If deletion succeeds the operator will remove `strimzi.io/topic-operator` from the `metadata.finalizer` and Kube will remove the resource.
+>        2. If deletion failed with `UNKNOWN_TOPIC_OR_PARTITION` (i.e. the topic didn't exist) the operator will continue to remove the finalizer. I.e. this case is not treated as an error.
+>        3. If deletion failed with any other error this is reported via a condition in the `status.conditions` and the finalizer is not removed.
+---
+
+Similarly to how it is possible to "manage" an existing unmanaged topic in Kafka, we will also provide a mechanism for "unmanaging" a managed topic, via a `spec.managed` flag.
+This will allow a `KafkaTopic` to be deleted **without** the topic in Kafka being deleted.
+This can be useful operationally, for example to change the `metadata.name` of a `KafkaTopic`.
```

Comment:
Have you considered removing the finalizer from a topic with `spec.managed: false`?

Reply:
Yes. If a CR with `spec.managed: false` gets deleted then the finalizer would be removed by the operator.

050-states.png shows all the different paths between the different states of the CR.

Note that, as currently specified, there is path where a CR can get deleted without the finalizer being removed by the operator. It happens if the operator's label selector initially matches the CR (so that the finalizer gets added), but the CR is then changed so that it no longer matches the operator's selector (the operator interprets this as "it's not my resource any more", assuming some other operator instance is now managing the CR. But if no other operator takes ownership then when the CR is then deleted there is nothing to delete the finalizer.

Reply:
> Yes. If a CR with spec.managed: false gets deleted then the finalizer would be removed by the operator.

What I meant here was if you considered removing the finalizer already when the `spec.managed` is set to `false` and not just when it is deleted.

> Note that, as currently specified, there is path where a CR can get deleted without the finalizer being removed by the operator. It happens if the operator's label selector initially matches the CR (so that the finalizer gets added), but the CR is then changed so that it no longer matches the operator's selector (the operator interprets this as "it's not my resource any more", assuming some other operator instance is now managing the CR. But if no other operator takes ownership then when the CR is then deleted there is nothing to delete the finalizer.

This would actually mean such resource cannot be deleted rather then `can get deleted without the finalizer being removed` as the finalizer might stay there. When the labels change and no longer match the selector, the TO will get a DELETE event. So I wonder if we should unset the finalizer at this event. But I guess that might raise conflicts with other instances and would not guarantee anything anyway. So maybe we have to live with it.